File: ios/Printer.swift
import Foundation

@objc(Printer)
class Printer: NSObject {
  @objc(print:withResolver:withRejecter:)
  func print(printActions: NSArray, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
    resolve(true)
  }
    
  @objc(open:withRejecter:)
  func open(resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
    resolve(true)
  }

  @objc(close:withRejecter:)
  func close(resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
    resolve(true)
  }
    
  // TODO: review this code, ChatGPT wrote it
  @objc(printHtml:withResolver:withRejecter:)
  func printHTML(printActions: NSArray, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    var html = ""

    for action in printActions {
      if let actionDict = action as? [AnyHashable: Any], let type = actionDict["type"] as? String {
        switch type {
        case "logo":
          if let data = actionDict["data"] as? String {
            let logoOptions = PrintAction.LogoOptions(options: actionDict["options"] as? [AnyHashable: Any])
            let printAction = PrintAction.Logo(data: data, options: logoOptions)
            
            let align: String
            switch printAction.options?.align {
              case .center:
                align = "center"
              case .right:
                align = "right"
              default:
                align = "center"
            }

            let data = printAction.data

            html += "<div style=\"text-align: \(align)\"><img src=\"data:image/png;base64, \(data)\"/></div>"
          }
        case "newLine":
          if let data = actionDict["data"] as? Int {
            let printAction = PrintAction.NewLine(data: data)
            
            let data = printAction.data
            html += "<div style=\"height: \(data * 10)px;\"></div>"

          }
        case "text":
          if let data = actionDict["data"] as? String, let optionsDict = actionDict["options"] as? [AnyHashable: Any] {
            let textOptions = PrintAction.TextOptions(options: optionsDict)
            let printAction = PrintAction.Text(data: data, options: textOptions)
            
            let align: String
            switch printAction.options.align {
              case .center:
                align = "center"
              case .right:
                align = "right"
              default:
                align = "center"
            }

            let size = printAction.options.size
            let fontWeight: String
            switch printAction.options.fontWeight {
              case .bold:
                fontWeight = "bold"
              default:
                fontWeight = "normal"
            }

            let data = printAction.data
            
            
            html += "<div style=\"text-align: \(align)\"><span style=\"font-size: \(size)px; font-weight: \(fontWeight)\">\(data)</span></div>"
          }
        default:
          break
        }
      }
    }

    resolve(html)
  }
}

File: android/src/main/java/com/pos/Printer.kt
package com.pos

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap

abstract class Printer {
  sealed class PrintAction {
    data class Logo(val data: String, val options: LogoOptions? = null) : PrintAction()
    data class NewLine(val data: Int) : PrintAction()
    data class Text(val data: String, val options: TextOptions) : PrintAction()

    data class LogoOptions(
      var align: Align = Align.LEFT,
    ) {
      constructor(options: HashMap<*, *>?): this(Align.LEFT) {
        val align = options?.get("align") as String?

        when (align) {
          "left" -> {
            this.align = Align.LEFT
          }
          "center" -> {
            this.align = Align.CENTER
          }
          "right" -> {
            this.align = Align.RIGHT
          }
        }
      }
    }

    data class TextOptions(
      var size: Int,
      var fontWeight: FontWeight = FontWeight.NORMAL,
      var align: Align = Align.LEFT,
    ) {
      constructor(options: HashMap<*, *>): this(12, FontWeight.NORMAL, Align.LEFT) {
        val size = options["size"] as Double
        val fontWeight = options["fontWeight"] as String?
        val align = options["align"] as String?

        this.size = size.toInt()

        when (fontWeight) {
          "normal" -> {
            this.fontWeight = FontWeight.NORMAL
          }
          "bold" -> {
            this.fontWeight = FontWeight.BOLD
          }
        }

        when (align) {
          "left" -> {
            this.align = Align.LEFT
          }
          "center" -> {
            this.align = Align.CENTER
          }
          "right" -> {
            this.align = Align.RIGHT
          }
        }
      }
    }

    enum class FontWeight {
      NORMAL,
      BOLD
    }

    enum class Align {
      LEFT, CENTER, RIGHT
    }
  }

  abstract fun print(actions: List<PrintAction>, promise: Promise)

  abstract fun open(promise: Promise)

  abstract fun close(promise: Promise)

  fun printHtml(printAction: PrintAction): String {
    var html = ""

    when (printAction) {
      is PrintAction.Logo -> {
        val align = when (printAction.options?.align) {
          PrintAction.Align.CENTER -> "center"
          PrintAction.Align.RIGHT -> "right"
          else -> "center"
        }

        val data = printAction.data

        html = "<div style=\"text-align: ${align}\"><img src=\"data:image/png;base64, ${data}\"/></div>"
      }
      is PrintAction.Text -> {
        val align = when (printAction.options.align) {
          PrintAction.Align.CENTER -> "center"
          PrintAction.Align.RIGHT -> "right"
          else -> "center"
        }

        val size = printAction.options.size
        val fontWeight = when (printAction.options.fontWeight) {
          PrintAction.FontWeight.BOLD -> "bold"
          else -> "normal"
        }

        val data = printAction.data

        html = "<div style=\"text-align: $align\"><span style=\"font-size: ${size}px; font-weight: ${fontWeight}\">$data</span></div>"
      }
      is PrintAction.NewLine -> {
        val data = printAction.data
        html += "<div style=\"height: ${data * 10}px;\"></div>"
      }
    }

    return html
  }
}
